[testharness.js] introduce `assert_precondition` (#19993)
web-platform-tests/wpt: docs/writing-tests/testharness-api.md

these places won't be detected correctly by the harness and may cause
unexpected exceptions that will lead to an error in the harness.

## Preconditions ##

> When a test would not be valid if certain conditions aren't met, but yet

Reviewer: This double negative is a bit hard to read;
```suggestion
When a test would be invalid unless certain conditions are met, but yet
```

doesn't explicitly depend on those preconditions, `assert_precondition` can be
used. For example:

```js
async_test((t) => {
  const video = document.createElement("video");
  assert_precondition(video.canPlayType("video/webm"));
  video.src = "multitrack.webm";
  // test something specific to multiple audio tracks in a WebM container
  t.done();
}, "WebM with multiple audio tracks");
```

A failing `assert_precondition` call is reported as a status of
`PRECONDITION_FAILED` for the subtest.

`assert_precondition` can also be used during test setup. For example:

```js
setup(() => {
  assert_precondition("onfoo" in document.body, "'foo' event supported");
});
async_test(() => { /* test #1 waiting for "foo" event */ });
async_test(() => { /* test #2 waiting for "foo" event */ });
```

A failing `assert_precondition` during setup is reported as a status of
`PRECONDITION_FAILED` for the test, and the subtests will not run.

> See also the `.optional` [file name convention](file-names), which is

Author: ```suggestion
See also the `.optional` [file name convention](file-names.md), which is
```

This is like https://github.com/web-platform-tests/wpt/pull/20038.
